torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -222,14 +223,20 @@ void DistEngine::executeSendFunction(
 
     // Run the autograd engine.
     runEngineAndAccumulateGradients(autogradContext, dummyRoot, outputEdges);
-
-    // Wait for all of the outstanding rpcs to complete.
-    autogradContext->clearAndWaitForOutstandingRpcs();
+    guard.cancel();
+
+    // Return future waiting for all of the outstanding rpcs to complete.
+    auto future = autogradContext->clearAndWaitForOutstandingRpcsAsync();
+    future->addCallback([contextId](const rpc::Message&) {
+      ClearContextIdGuard guard2(contextId);
```

Comment:
I would turn this into an explicit call to clear the context ID. The RAII wrapper does the same on destruction, which in this case is one level of indirection too many in my opinion. If it's too much code or duplication, we could consider adding a helper to clear the context ID of the current distributed autograd instance. cc @pritamdamania87

Reply:
Right, I agree. Actually, I rendered this section back to the original.

At this point, my thinking is, the right thing is to clear for the current calling thread (as originally), and if the handler is in fact async, for that specific handler to transfer the thread_local context to the async thread. If there's a generalizable pattern with respect to this, we should generalize it when it becomes more clear.